### PR TITLE
Build path info directly instead of loading REST API.

### PR DESCRIPTION
**Pull request by vedanshujain**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

While loading blocks, we also load the entire REST API as part of the Store API's initialization since they are powered by store API. We also hydrate responses from several store API endpoints in the first page load itself to prevent having to make requests multiple times.

Unfortunately, when loading REST API, we load the entirety of all REST API available on a site, and not just the controllers required to power blocks. This can cause a significant performance degradation especially if the site has few plugins that provide extensive REST API support (such as WooCommerce). 

With WC 8.5.x we are now loading blocks with many themes by default (via the mini cart block), and this has exacerbated problems for many sites because of this extra loading.

In this PR, we are going to skip loading REST API both for building routes as well as for data hydration. Instead, we will call the controller directly which we need to build our response array for the mini cart (and other blocks). We do this, by introducing a function that can find the controller that we need based on the regex that the controller provides and then matching it against the API path that we want to load.

In my testing, this prevents loading REST API and provides significant performance improvements, especially on sites that have lots of plugins enabled which provide REST API routes.

Towards #43879

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR needs exploration testing of blocks features, I recommend exploration testing following blocks:

1. Mini cart block.
2. Cart and checkout block (perform a checkout entirely using WC blocks).
3. Product list and filters blocks.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/StoreApi/RoutesController.php [hunk 1]
```diff
@@ -104,6 +111,35 @@ public function get( $name, $version = 'v1' ) {
...
+	 * @param string $controller Whether to return controller name. If false, returns empty array. Note:
+	 * When $controller param is true, the output should not be used directly in front-end code, to prevent class names from leaking. It's not a security issue necessarily, but it's not a good practice.
+	 * When $controller param is false, it currently returns and empty array. But it can be modified in future to return include more details about the route info that can be used in frontend.
+	 *
+	 * @return string[] List of route paths.
+	 */
+	public function get_all_routes( $version = 'v1', $controller = false ) {
+		$routes = [];
+
+		foreach ( $this->routes[ $version ] as $key => $route_class ) {
```
Comment (vedanshujain) [inline plugins/woocommerce/src/StoreApi/RoutesController.php:129]: Can an external plugin add their endpoint to the `$this->routes` array? AFAICS they can't, but if it's possible, then we would need to drop the `method_exists` check from line 131.

File: plugins/woocommerce/src/Blocks/Domain/Services/Hydration.php [hunk 2]
```diff
@@ -38,9 +41,37 @@ public function __construct( AssetDataRegistry $asset_data_registry ) {
...
+			);
+
+			$response = $controller->get_response( $request );
+			return array(
+				'body'    => $response->get_data(),
+				'headers' => $response->get_headers(),
+			);
+		}
+
+		$this->cache_store_notices();
```
Comment (mikejolley) [inline plugins/woocommerce/src/Blocks/Domain/Services/Hydration.php]: This needs to be ran before any controllers are called—this is so endpoints like checkout have a clean slate in case notices are added during payment.

File: plugins/woocommerce/src/StoreApi/RoutesController.php (hunk 1 above)
Comment (mikejolley) [reply plugins/woocommerce/src/StoreApi/RoutesController.php:129]: No we didn't make this part extensible.

File: plugins/woocommerce/src/StoreApi/RoutesController.php [hunk 3]
```diff
@@ -104,6 +111,35 @@ public function get( $name, $version = 'v1' ) {
...
+
+			if ( ! method_exists( $route_class, 'get_path_regex' ) ) {
```
Comment (mikejolley) [inline plugins/woocommerce/src/StoreApi/RoutesController.php:131]: What was the reason for the new method vs `get_path`?

**Review by mikejolley — CHANGES_REQUESTED** [review]
This does appear to shave a few seconds off so excellent. I tested some direct rest API calls and the hydration and things worked as normal.

There is a change needed regarding the notice caching.

@opr or @senadir would you take a look also? Is there anything on the extensibility side you could see breaking?

File: plugins/woocommerce/src/StoreApi/RoutesController.php (hunk 3 above)
Comment (senadir) [reply plugins/woocommerce/src/StoreApi/RoutesController.php:131]: Same question, I think to reflect that `get_path_regex` will be used in a regex check vs `get_path` which should return a static fully qualifying string?

File: plugins/woocommerce/src/Blocks/Domain/Services/Hydration.php (hunk 2 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/Blocks/Domain/Services/Hydration.php]: Added here https://github.com/woocommerce/woocommerce/pull/44080/commits/8eed376ff294b724863a7954e9c2682b3468078a

File: plugins/woocommerce/src/StoreApi/RoutesController.php (hunk 3 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/StoreApi/RoutesController.php:131]: The primary reason was that `get_path` is an instance method, so we would need to create an instance of it to build our paths array. On the other hand, `get_path_regex` is static, so we can build our path array without having to create an instance, which is usually cheap, but can get expensive when doing a lot of it.

**Comment by vedanshujain** [comment]
@mikejolley / @senadir Ready for another round (CI failures looks unrelated)

**Review by mikejolley — APPROVED** [review]
Tests are passing now. I tested:

- Cart and checkout flow, while watching network tab for requests
- Hydration visible on page
- Sorting/browsing shop page blocks
- Direct calls to the API (v1 and without v1)

I notice a request on every page from mini cart but this is present in trunk as well so seems unrelated to this work. Based on that and the timings, approving 👍🏻 

@senadir was also looking so await his approval also.

File: plugins/woocommerce/src/StoreApi/StoreApi.php [hunk 4]
```diff
@@ -35,6 +35,24 @@ function() {
...
+				$routes = array_merge(
+					$routes,
+					self::container()->get( RoutesController::class )->get_all_routes( 'v1' )
+				);
+
+				return $routes;
+			},
+			10,
+			3
+		);
```
Comment (senadir) [inline plugins/woocommerce/src/StoreApi/StoreApi.php:55]: It's not explained anywhere why is this needed.

**Review by senadir — APPROVED** [review]
This is testing fine for me as well.

**Comment by vedanshujain** [comment]
Closing and reopening to trigger CI

File: plugins/woocommerce/src/StoreApi/StoreApi.php (hunk 4 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/StoreApi/StoreApi.php:55]: My bad, it's buried in the wall of text in description. So this adds a hook 
 [in `woocommerce_blocks_pre_get_routes_from_namespace` ](https://github.com/woocommerce/woocommerce/pull/44080/files/bc5d51b0208f8609f2a180fa82b63f1b15ac3b56#diff-aab09b3f15d65e08d6da3daf65c24daa0564dee14773891f401ccf2683d761b5R474) and makes use of the opportunity to provide a pre-built routes array without loading the REST API. If we don't hook here, than the corresponding `apply_filter` will end up loading the REST API to build store API routes.
